@xen-orchestra/audit-core/src/index.js

```diff
@@ -0,0 +1,121 @@
+// see https://github.com/babel/babel/issues/8450
+import 'core-js/features/symbol/async-iterator'
+
+import assert from 'assert'
+import hash from 'object-hash'
+import { invert } from 'lodash'
+
+export class Storage {
+  constructor() {
+    this.locked = undefined
+  }
+
+  async lock() {
+    if (this.locked !== undefined) {
+      await this.locked
+    }
+
+    let res
+    this.locked = new Promise(resolve => {
+      res = resolve
+    })
+    return () => {
+      this.locked = undefined
+      return res()
+    }
+  }
+}
+
+// Format: $<algorithm>$<salt>$<encrypted>
+//
+// http://man7.org/linux/man-pages/man3/crypt.3.html#NOTES
+const ALGORITHM_TO_ID = {
+  sha256: '5',
+}
+
+const ID_TO_ALGORITHM = invert(ALGORITHM_TO_ID)
+
+export const NULL_ID = 'nullId'
+
+const HASH_ALGORITHM = 'sha256'
+const createHash = (data, algorithm = HASH_ALGORITHM) =>
+  `$${ALGORITHM_TO_ID[algorithm]}$$${hash(data, {
+    algorithm,
+    excludeKeys: key => key === 'id',
+  })}`
+
+export class AuditCore {
+  constructor(storage) {
+    assert(storage !== undefined && storage.lock !== undefined)
```

Comment:
The `AuditCore` implementation requires lock method to be defined see https://github.com/vatesfr/xen-orchestra/pull/4733/files/32385ce0b2d3d5e1f2408a408c899eccab3e27b9#diff-e8dd74e1644cd1cb6f4d7143b0506242R55

Reply:
And it will be always defined because storage must extend `Storage`, therefore the test is unnecessary.

Reply:
I thought you were checking `storage.locked` (instead of `storage.lock`) so ignore my comment but I'll leave the conversation open for @julien-f's comment, though.